### ISSUE TITLE
Co-Op

### ISSUE THREAD

**Issue by krakjoe** [krakjoe/parallel#25]
I've implemented the beginnings of coop ...

Notes:

Co-Op in parallel allows you to effectively achieve 1:N threading, by yielding to the scheduler and allowing it to run any waiting tasks on the stack, optionally re-scheduling the current task.

I've chosen the method name yield, but it may not be a good name: yield in PHP means a very specific thing today, and yielding to the scheduler is somewhat different ...

This should probably be marked experimental in the first release that includes it ...

TODO
---------

**Important**:
 - [x] native frames
 - [x] yield at instruction (closer to native generators)
 - [x] yielding values
 - [x] Cleanup code

**Less important**:
 - [ ] Discover how well supported pthread_yield is, and how practical it is to use it

**Comment by ghost** [comment]
Do you think it makes sense to make this part of the parallel extension? There is ext-async which already provides task scheduling and async parts (e.g. file I/O) for single-threaded PHP.

IMO it makes sense to work together with martinschroeder to make ext-async compatible with ext-parallel (I didn't try ext-async with ext-parallel yet), as ext-async already provides powerful features and reimplementing this into ext-parallel doesn't make much sense to me, if ext-async can be (or is?) made compatible/safe to use within ext-parallel. But I don't have the required knowledge to know that, but *if* it's possible without any troublesome amount of effort, then I think it's worthwhile to pursue that instead of reimplementing parts of it.

Link: https://github.com/concurrent-php/ext-async

**Comment by krakjoe** [comment]
> Do you think it makes sense to make this part of the parallel extension?

Yes, clearly :)

(EDIT: I meant, "clearly I think it makes sense because I'm spending time on it", not "clearly it makes sense because it so obviously makes sense" ...)

> There is ext-async which already provides ...

There are many projects that do the same ... to be absolutely clear, the aim here is not to introduce a new standard or render obsolete any of those projects, but to enable that kind of project to interact with the scheduler for parallel. Today, yield is prohibited directly in tasks because it's broken (or doesn't make sense) - there is nothing to yield too (or may not be). What `Runtime::yield` does is replace the broken yield instruction directly in the task. That's as far as it goes, in this extension ....

**Comment by krakjoe** [comment]
@trowski @kelunik @bwoebi can I get some input from you guys here please ?

I wanted originally to use built in generator for yielding, just restoring the functionality for the top most context (the function that is passed to ::run) ... but the built in generator very much wants to yield to an outer context that doesn't exist, and I don't seem to be able to intercept (even with setting user opcodes, which I don't want to do) at the right time to change the behaviour of built in generator.

So I've gone with a method, and you can yield a value and control to the scheduler, and the next "generator" (function that yields) may receive the last yielded value if it uses the return value ...

Because we just return the value, and not an iterable object like a builtin generator, it makes code look a bit awkward ... here's some example code:

```php
<?php
use \parallel\Runtime;

$parallel = new Runtime();

$parallel->run(function(){
    echo "[A] enter\n";
    while ($i++<10) {

        
        echo "[A] suspend\n";
        Runtime::yield([$i]);
        echo "[A] resume\n";
    }
    echo "[A] leave\n";
});

$parallel->run(function(){
    echo "[B] enter\n";
    echo "[B] suspend\n";
    if (($values = Runtime::yield())) {
        do {
            echo "[B] resume\n";
            foreach ($values as $value) {
                printf("## in b %d\n", $value);
                /* Runtime::yield(); optionally yield during iteration */
            }
            echo "[B] suspend\n";
        } while ($values = Runtime::yield());
    }
    echo "[B] leave\n";
});

$parallel->run(function(){
    echo "[C] enter\n";
    while ($a++<15) {
        echo "## in c\n";
        
        echo "[C] suspend\n";
        Runtime::yield();
        echo "[C] resume\n";
    }
    echo "[C] leave\n";
});
?>
```

Just looking for thoughts on any of this ....

EDIT: defer that feedback ... I think we can have native (looking) generators ...

**Comment by kelunik** [comment]
@krakjoe Your example doesn't make much sense to me that way. It works with exactly two threads, but if there's a third one, which one will receive the values?

I don't think we need `Runtime::yield()`. `ext-async` will take care of green threads, while `ext-parallel` should focus on "real" threads, IMO.

**Comment by krakjoe** [comment]
Hmm ... tasks that assign the result of yield() to a value receive the last value passed as the first argument of yield() ... that seemed to make sense, no ?

I'm not wedded to the idea of yielding values at all, you can use a channel to pass values, but yielding doesn't have the object/resource restrictions that channels have ...

If the values don't or can't make sense, what about just yielding to the scheduler, so we can fix yield instruction, that seems to be the minimum to me ?

**Comment by krakjoe** [comment]
I've dropped coop from the develop branch for now and moved it into it's own branch ... I'll await feedback from more people who would actually be using it and go back to channels for now, as that seems more important ...

**Comment by krakjoe** [comment]
It would be nice to hear other peoples thoughts on this too ... ping @everyone

**Comment by ghost** [comment]
I have tested using `ext-parallel` together with `ext-async` and it works to some degree. One can use threads to run multiple event loops (async task schedulers that is) within the same PHP process and it works just fine. The problem (from an `async` or non-blocking point of view) is not related to threads (as they just work) but to the integration of futures and channels. They need to block a thread while they wait for other threads to produce a result value and thereby block the event loop during wait.

I created a new branch called `threads` within `ext-async` and basically recreated your `Runtime` API in a way that works well with `libuv`. The basic pattern i used is to have threads wait for new jobs using the same mutex / condition wait approach that you are using. It differs in the way threads notify the main process of completion: it uses something like [eventfd](http://man7.org/linux/man-pages/man2/eventfd.2.html) and a queue of finished jobs protected by a mutex.

Maybe you can have a look into [the source](https://github.com/concurrent-php/ext-async/blob/threads/src/thread.c) and check if i screwed it up too much. ;-) AFAIK code that uses event loops needs some kind of file descriptor approach to poll for completion instead of blocking the thread during wait but there might be other ways to achieve integration in a better way.

I really like your work on this new extension, nice and easy to use, way better than pthreads. :smiley:

**Comment by krakjoe** [comment]
Hi Martin :)

I think the answer to this problem is probably not copy pasta :) 

I've looked at ext-async, but not in very great depth, except the last twenty minutes reading the threads branch ...

When you introduce threads into your async extension you complicate it almost infinitely: How do tasks in separate threads share your implementation of a Channel or any other complex type ? 

Your implementation of Channel has:

```
    public function isReadyForSend(): bool { }
    
    public function isReadyForReceive(): bool { }
```

If you could share this complex type among threads, it would encourage the user to write code with race conditions, you can't have those methods (or that kind of logic) while multiple threads may be manipulating the same channel concurrently.

> AFAIK code that uses event loops needs some kind of file descriptor approach to poll for completion instead of blocking the thread during wait but there might be other ways to achieve integration in a better way.

Look at poll.c in this repo: You don't necessarily need a file descriptor to poll for state, you can poll whatever you like, it's just normally involving nbio and so streams. You can think of a channel or future as a pseudo kind of stream and that is how Events in parallel treats them.

I view parallel and asynchronous concurrency as very different strategies, for very different things. That's not to say that there is absolutely no overlap, or that I object to some kind of integration of the two.

But what kind of integration ? What do we actually want to achieve ? What is reasonably achievable ?

**Comment by ghost** [comment]
Just a little information: Rust has "multi producer, single consumer" channels, that means the sender can be shared (cloned) with multiple threads. https://doc.rust-lang.org/1.5.0/std/sync/mpsc/index.html

However I don't think that Channels here need to do that. Simply one producer and consumer is enough in my opinion.

**Comment by ghost** [comment]
Hey Joe :notes:,

i do agree with you that this is not a case for copy-paste and that threads and fibers are very different in the way they work and it is not easy (sometimes not even possible) to combine them in a reasonable way.

> When you introduce threads into your async extension you complicate it almost infinitely: How do tasks in separate threads share your implementation of a Channel or any other complex type ?

I was thinking a lot more "shared-nothing" here and only thought of threads to offload CPU intensive computations to a worker and grab the result without blocking the event loop. Passing any kind of object backed by `libuv` to another thread is not supported anyways because the loop can only be operated from a single thread. The only exception seems to be `uv_async_t` which is threadsafe and needed to notify the event loop in another thread. Only primitive types and arrays may be exchanged between threads in this setup which is basically what could be JSON-encoded and passed between processes using IPC pipes as well.

I thought about adding a specialized implementation of `Channel` to facilitate yielding multiple values from a thread (or passing them into a thread) but you can always do the same by feeding results from threaded jobs into a channel using an async task.

> But what kind of integration ? What do we actually want to achieve ? What is reasonably achievable ?

I think that exposing a file descriptor that can be polled by an event loop to wake the loop up whenever thread state has changed would be the most straightforward integration path (and could by used by @WyriHaximus to avoid the need for polling using a periodic timer). Maybe this could be added to your `Events` class as an additional API complementing `poll()`?

**Comment by trowski** [comment]
> I think that exposing a file descriptor that can be polled by an event loop to wake the loop up whenever thread state has changed would be the most straightforward integration path

+1 if possible. This would allow an immediate reaction to a thread changing state instead of checking every X milliseconds.

In `amphp/parallel` I don't use the Channel API. Instead I use sockets so they can polled by the event loop. Each socket only goes between the parent and single child, so I don't have the thread synchronization problems that Channels here do. Though I also don't have the flexibility either.

**Comment by WyriHaximus** [comment]
> I think that exposing a file descriptor that can be polled by an event loop to wake the loop up whenever thread state has changed would be the most straightforward integration path (and could by used by @WyriHaximus to avoid the need for polling using a periodic timer). Maybe this could be added to your `Events` class as an additional API complementing `poll()`?

That would be welcome, or alternatively a callback that gets triggered when new data comes in

**Comment by krakjoe** [comment]
There is no file descriptor I could expose, nor is there a way to emulate a file descriptor (that I'm aware of).

**Comment by dktapps** [comment]
what about eventfd() ?

**Comment by ghost** [comment]
For Windows the event objects can probably be used (they're compatible with the win32 implementation of PHP's `select` (`WaitForMultipleObjects`)). <https://docs.microsoft.com/en-us/windows/desktop/sync/event-objects>

**Comment by krakjoe** [comment]
Even so, exposing an emulated file descriptor won't do it, you cannot write to channels in an unsafe loop ... 

Some thought dumps ...

I guess it's just not clear to me what we're actually trying to achieve here ... the "co-op" support I talked about here was just "give tasks ability to cooperate with scheduler in parallel", and now we're talking about integrating with event loops, but in order to do what ? When none of the objects that any event loop implements (be it extension or userland code) can be shared among threads and are not thread safe by design even if they could be shared, what are we going to achieve with this kind of integration ?

When the subject of integration was first mentioned, I imagined it would be a case of the parallel scheduler yielding to another scheduler, and or providing a way for ext-async to access the parallel scheduler in some way. This makes some sense to me, it doesn't matter that no userland or extension ev loop API is thread safe, because a scheduler is thread local anyway ... 

It might even make sense to have an Events::addSource(EventInterface $s) - Events is also thread local in that you can't pass Events among threads, but it's thread safe, in that multiple threads may be watching the same channels ... writes would still have to be orchestrated by the parallel ev loop, and so doesn't really fit normal I/O, it would also probably need to include exclusivity, but I haven't thought about this very much ...

The kind of errors that are permissible for async I/O (EAGAIN etc) are not permissable in parallel, that's why the Events loop needs an Input, I don't know how to make it look like I/O while keeping it safe, I don't think it's possible ...

**Comment by ghost** [comment]
Here are some of my thoughts on this (unfortunately off-topic as well in terms of what you meant by "co-op"). The basic idea i had was that it night be possible to do something like this:
```php
$events = new Events();
$events->setInput($input = new Input());
$events->setTimeout(0);

$loop->addReadStream($events->getPollFd(), function ($fd) use ($loop, $events, $input) {
    try {
        foreach ($events as $event) {
            // Do somethin with event.
        }
    } catch (Timeout $e) {
        if (count($input) == 0 && count($events) == 0) {
            $loop->removeStream($fd);
        }
    }
});
```
Now assuming that "poll fd" becomes readable whenever one of the input values has been consumed by a channel or a value was read from a channel this would integrate nicely with an application based on an event loop. The timeout value of `0` should avoid blocking the thread during poll. The callback might read no event (in case another thread was faster) or a sequence of one or more events.

A channel will be added to `$events` if we want to read values. Adding a value to `$input` would be needed to write the value. I do not know if this can be achieved but it seemed like the easiest integration path to me. Of course one can always use a periodic timer with an event loop and call `poll()` with a timeout of `0` to achieve something similar, but this will lead to a very "busy" event loop and a lot of polling that results in timeout exception.

> and now we're talking about integrating with event loops, but in order to do what ?

In order to make use of the best of both worlds. Event loops are very good at scaling (non-blocking) IO while threads allow us to make good use of multiple CPU cores to do computations that would block the event loop and increase IO latency. There are other ways to do this (multiple processes) but with threads it just seems to be more efficient and straightforward. If you see any way for such an integration maybe sometime you could open another issue and ping me as i would be very much interested in this.  :smiley:

**Comment by TheTechsTech** [comment]
> and yielding to the scheduler is somewhat different ...

Have you ever reviewed **Nikita Popov** [Cooperative multitasking using coroutines (in PHP!)](https://nikic.github.io/2012/12/22/Cooperative-multitasking-using-coroutines-in-PHP.html).

You might see that he show it to be an different opinion. I released something similar to what's here that works in userland, with no extensions, https://symplely.github.io/coroutine/. 

I'm hoping to see some examples so i can find an easy way to integrate with your library.

**Comment by krakjoe** [comment]
> The PHP Internals team has recently released Parallel.

Internals doesn't have anything to do with parallel, it's not an official extension ...

> however this Coroutine package does everything listed in there

That seems wildly inaccurate.

>  Parallel concurrency API without the unnecessary restrictions, PHP 7.2, and limits on Tasks

There are no *unnecessary* restrictions or limits, the restrictions that do exist, exist to keep code safe.

> I'm hoping to see some examples so i can find an easy way to integrate with your library.

For the moment, I don't intend to provide lower level integration for fibers, coroutines, or any other kind of asynchronous processing. parallel and asynchronous concurrency have different domains, I'm focusing on the domain that is relevant to this extension.

That's not to say that there can't be some kind of integration, but at this time it's not a priority, and it may never become a priority.

I'm going to close this thread for now, because I don't intend to work on it.

**Comment by TheTechsTech** [comment]
> > The PHP Internals team has recently released Parallel.
> 
> Internals doesn't have anything to do with parallel, it's not an official extension ...

Well, just figured since it's on https://phpinternals.news/11 the podcast and there website, that it is. 
I think most will come to that conclusion. 
I know nothing of internals process on that.

> > however this Coroutine package does everything listed in there
> 
> That seems wildly inaccurate.

I think you feeling infringed on with that statement. 
Only the code would show otherwise. My work is not fully done. All my current examples are others, and  from other languages.Anyway, if one has any real understand of programming, 
I mean basic computer science concepts, one could solve any problem, using in any language. 
Might be a long way with some languages, but it would prove to oneself they actually grasp the concept. 
The wording will become irrelevant, the results will be the same.

> For the moment, I don't intend to provide lower level integration for fibers, coroutines, or any other kind of asynchronous processing.

Really don't what you mean by 'provide lower level', i wanted to see something more clear cut that the tests. 
I can recreate them with my library with the parallel class i use, which is an wrapper around symfony process class. I was just hoping to drop it, if there is an real speed benefit or simpler way to get things done. **Maybe the only thing i like, is that, this package is available on windows.**

**Comment by WyriHaximus** [comment]
> **Maybe the only thing i like, is that, this package is available on windows.**

Well tbh either don't use windows or use Docker/Vagrant/WSL on windows

**Comment by ClosetGeek-Git** [comment]
My own brain dump: you've in effect created a worker pool that is able to handle jobs much like JS callbacks which are often chained to create a somewhat asynchronous chain of execution within a threadpool as well. Providing non-iterative yield within PHP\Parallel::yield would then be the functional equivalent to a JS generator function* which too have the same behavior but without iterators such as with standard javascript generators. Also note that Python can also call yield as a first-class return type without using generators or other iterative objects.

In other words the behavior your suggesting has an almost identical equivalent in Javascript, and basic non iterative yields are standard in python.

But no, this isn't the exact behavior as PHP Generator objects which extend the PHP Iterator class, but I don't see why it should have to. In fact I found iterating a generator annoying and somewhat confusing when I first approached the PHP generator object for co-op.

**Comment by TheTechsTech** [comment]
You are getting the gist of it. You could study the origin of Python usage of `yield`, and how it was used to to achieve `await`. What you will find is that, it's just a different keyword usage, syntax sugar. 

> But no, this isn't the exact behavior as PHP Generator objects which extend the PHP Iterator class, but I don't see why it should have to. In fact I found iterating a generator annoying and somewhat confusing when I first approached the PHP generator object for co-op. 

Pythons native asynchronous behavior is actually all Python, not `C code`, not a extension, and deep down there code base `yield` is still what makes it work. 

Unfortunately, not many average PHP developers follow the concepts or understand much computer science studies.
